### PR TITLE
Avoid putting -repo_config after directories

### PR DESCRIPTION
**Pull request by linzhp**
**What type of PR is this?**
Bug fix

**What package or component does this PR mostly affect?**

gazelle.bash

**What does this PR do? Why is it needed?**
Gazelle can take directories as positional arguments. When gazelle.bash append `-repo_config` at the end of the arguments, the flag itself will be treated as directories if there are positional arguments in the arguments, leading to errors like:

```
gazelle: -repo_config: failed to resolve symlinks: lstat /path/to/repo/root/-repo_config: no such file or directory
```

To fix it, we need to put `-repo_config` before positional arguments. However, because Gazelle takes subcommands like "fix", "update", we also need to put `-repo_config` after those commands. In order to do that, this PR reimplemented the logic of merging command line args with generated args in gazelle.bash, and make sure that the first argument is always a subcommand, so we can safely insert the `-repo_config` to the second position.

cc. @cgrindel @seh

**Other notes for review**
Is there a better way to test this than changing the presubmit.yml?

### CODE REVIEW

File: internal/gazelle.bash.in [hunk 1]
```diff
@@ -62,12 +62,18 @@ function set_goroot {
...
-  shift
-  ARGS+=("$@")
-elif [ $# -ne 0 ]; then
-  ARGS=("$@")
-fi
+case "${1-}" in
+  "fix" | "update" | "help" | "update-repos")
+    ARGS=("$@")
+    ;;
+  "-args")
```
Comment (linzhp) [inline internal/gazelle.bash.in:69]: @jayconrod do we need to support this case? It's undocumented and only used by autogazelle, which should be easy to update

Comment (jayconrod) [reply internal/gazelle.bash.in:69]: Should be fine to drop or change. I don't think it works very well now.

**Review by jayconrod — APPROVED** [review]

**Review by seh — APPROVED** [review]
Thank you for both explaining the problem so clearly and covering the cases in the patch.

Is Gazelle's flag parsing unusual, in that it's mistaking this command-line flag for an argument? Is that something that we can improve as well?

**Comment by linzhp** [comment]
> Is Gazelle's flag parsing unusual, in that it's mistaking this command-line flag for an argument?

Gazelle uses Go's [built-in flag package](https://pkg.go.dev/flag) package to parse. This is a known issue for that package. Some third-party libraries do a better job, but it's probably not a good idea to introduce a third-party library for this purpose.

**Comment by jayconrod** [comment]
Gazelle follows similar conventions to other Go command-line tools with subcommands, but those are different from most other languages. The structure is:

```
gazelle [subcommand] [flags] [arguments]
```

The subcommand may be `fix`, `update`, `update-repos`, or `help`. It's optional to avoid a breaking change since initially there weren't subcommands (`update` was the only behavior) and it was common to install and invoke Gazelle directly, not through a Bazel rule.

Gazelle takes the subcommand out of the argument list before parsing the flags since there are different flags for different subcommands.

After that though, Go's `flag` package requires that flags come before arguments. So adding a flag to the end of the list doesn't work.

**Comment by seh** [comment]
Understood. Thank you both for clarifying. I'm more used to _spf13/pflag_ (mostly by way of _spf13/cobra)_, which is more flexible.

**Pull request merged**